gc.c

```diff
@@ -1938,24 +1937,50 @@ object_id(VALUE obj)
     return object_id0(obj);
 }
 
+struct build_id2ref_args {
+    st_table *id2ref_tbl;
+    rb_objspace_t *objspace;
+};
+
 static void
 build_id2ref_i(VALUE obj, void *data)
 {
-    st_table *id2ref_tbl = (st_table *)data;
+    struct build_id2ref_args *args = (struct build_id2ref_args *)data;
+    st_table *id2ref_tbl = args->id2ref_tbl;
+    rb_objspace_t *objspace = args->objspace;
 
+    // If the object is garbage we must not insert it.
+    // But also we must clear its object id, otherwise when it will be finally freed
+    // obj_free_object_id will crash.
     switch (BUILTIN_TYPE(obj)) {
       case T_CLASS:
       case T_MODULE:
         if (RCLASS(obj)->object_id) {
-            st_insert(id2ref_tbl, RCLASS(obj)->object_id, obj);
+            if (rb_gc_impl_garbage_object_p(objspace, obj)) {
+                RCLASS(obj)->object_id = 0;
+            }
+            else {
+                st_insert(id2ref_tbl, RCLASS(obj)->object_id, obj);
+            }
         }
         break;
       case T_IMEMO:
+        if (IMEMO_TYPE_P(obj, imemo_fields) && rb_shape_obj_has_id(obj) && rb_gc_impl_garbage_object_p(objspace, obj)) {
+            rb_imemo_fields_clear(obj);
+        }
```

Comment:
A change I've meant to do for a while, and that may simplify all this.

I think it'd be helpful if `imemo_fields.klass` pointed to the owner object rather than its class. In situations like this, it would allow to deal with `object_id` entirely from the `T_IMEMO`, and we could ignore the associated object, breaking the interdependency.